core/src/main/java/io/stargate/core/activator/BaseActivator.java

```diff
@@ -0,0 +1,227 @@
+import org.osgi.framework.InvalidSyntaxException;
+import org.osgi.framework.ServiceReference;
+import org.osgi.util.tracker.ServiceTracker;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class BaseActivator implements BundleActivator {
+  private static final Logger logger = LoggerFactory.getLogger(BaseActivator.class);
+
+  private final String activatorName;
+
+  protected BundleContext context;
+
+  private Optional<Class<?>> targetServiceClass;
+
+  public boolean started;
+
+  @VisibleForTesting
+  @GuardedBy("this")
+  public Tracker tracker;
+
+  /**
+   * @param activatorName - The name used when logging the progress of registration.
+   * @param targetServiceClass - This class will be used when registering the service. If null, then
+   *     the registration will not happen.
+   */
+  public BaseActivator(String activatorName, @Nullable Class<?> targetServiceClass) {
+    this.activatorName = activatorName;
+    this.targetServiceClass = Optional.ofNullable(targetServiceClass);
+  }
+
+  /**
+   * Convenience method for activators that does not register any service see docs for {@link
+   * this#BaseActivator(String, Class)}.
+   */
+  public BaseActivator(String activatorName) {
+    this(activatorName, null);
+  }
+
+  /**
+   * When the OSGi invokes the start() method, it constructs the {@link Tracker}. It will listen for
+   * notification of all services passed as {@link this#dependencies()}. It will wait for a
+   * notification denoting that service was registered using {@link
+   * Tracker#addingService(ServiceReference)}. If all services are present, it will call the
+   * user-provided {@link this#createService()} and register it in the OSGi using {@link
+   * BundleContext#registerService(Class, Object, java.util.Dictionary)} if {@code
+   * targetServiceClass.isPresent()}. If it is not present, the {@link this#createService()} is
+   * called but there will bo no registration in the OSGi.
+   */
+  @Override
+  public synchronized void start(BundleContext context) throws InvalidSyntaxException {
+    logger.info("Starting {} ...", activatorName);
+    this.context = context;
+    tracker = new Tracker(context, context.createFilter(constructDependenciesFilter()));
+    tracker.open();
+  }
+
+  @VisibleForTesting
+  String constructDependenciesFilter() {
+    StringBuilder builder = new StringBuilder("(|");
+
+    for (ServicePointer<?> servicePointer : dependencies()) {
+      if (servicePointer.identifier.isPresent()) {
+        builder.append(String.format("(%s)", servicePointer.identifier.get()));
+      } else {
+        builder.append(String.format("(objectClass=%s)", servicePointer.className.getName()));
+      }
+    }
+    builder.append(")");
+    return builder.toString();
+  }
+
+  @Override
+  public synchronized void stop(BundleContext context) throws Exception {
+    if (started) {
+      logger.info("Stopping {}", activatorName);
+      stopService();
```

Comment:
We automatically register the service, but do not automatically unregister it... this looks a bit awkward to me.

How about changing the sub-class lifecycle from `createService/stopService` to `startService/stopService` and delegate registration to sub-classes? This should also maturally suppose use cases where the sub-class registers more than one service object (from the same set of dependencies), which may soon be the case on the `feature-sql` branch.

Reply:
The `Tracker` extends `ServiceTracker` and I think it handles the de-registration automatically:
```
	public void removedService(ServiceReference<S> reference, T service) {
		context.ungetService(reference);
	}
```

Reply:
Re: registering in the child classes. I think that hiding this logic internally is quite convenient because you don't need to worry about it nor test it. For the N services use case, maybe it would be better to just rework the `BaseActivator` logic for this use case?

Reply:
sgtm.

Reply:
I've added de-registration of target service: https://github.com/stargate/stargate/pull/370/commits/d39e9e92f95d2672bf78635957e1173536fa2767